sources/thread_pool.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef THREAD_POOL_HPP
+#define THREAD_POOL_HPP
+
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <queue>
+#include <atomic>
+#include <iostream>
+
+struct thread_pool {
+    using job = std::function<void()>;
+
+private:
+
+    std::vector<std::queue<job>>                          work_queues;
+    std::vector<std::unique_ptr<std::mutex>>              locks;
+    std::vector<std::unique_ptr<std::condition_variable>> condvars;
+    std::vector<std::unique_ptr<std::thread>>             threads;
+
+    std::mutex job_done_lock;
+    std::condition_variable job_done;
+
+    std::atomic<size_t> rr_ind;
+
+    std::atomic<size_t> active_count;
+
+    bool active;
+
+public:
+    thread_pool(size_t num_threads) {
+        active = true;
+        rr_ind.store(0);
+        active_count.store(0);
+
+        work_queues.resize(num_threads);
+        threads.resize(num_threads);
+        locks.resize(num_threads);
+        condvars.resize(num_threads);
+
+        for(size_t i=0; i < num_threads; ++i) {
+            locks[i].reset(new std::mutex());
+            condvars[i].reset(new std::condition_variable());
+
+            threads[i].reset(new std::thread(
+                        [i,this]() {
+                while(active) {
+                    job f = []() {};
+                    {
+                        std::unique_lock<std::mutex> l(*locks[i]);
+
+                        while(work_queues[i].empty() && active) {
+                            condvars[i]->wait(l);
+                        }
+                        if(!active) { break; }
+                        ++active_count;
+
+                        f = work_queues[i].front();
+                        work_queues[i].pop();
+                        /* std::cout << "Thread " << i << ": tick\n"; */
+                    }
+                    f();
+                    --active_count;
+                    job_done.notify_all();
+                }
+            }));
+
+        }
+    }
+
+    ~thread_pool() {
+        active = false;
+        for(auto& cv: condvars) {
+            cv->notify_all();
+        }
+        for(auto& t: threads) {
+            t->join();
+        }
+    }
+
+    void add_job(job j) {
+        size_t i = (rr_ind++)%threads.size();
+        /* std::cout << "Adding to queue " << i << std::endl; */
+        {
+            std::lock_guard<std::mutex> l(*locks[i]);
+
+            work_queues[i].push(j);
+        }
+        condvars[i]->notify_all();
+    }
+
+    bool done() {
```

Comment:
I have a hard time following what `done()` and `wait()` does, or what is the use of `condvars`. Besides, `done()` seems susceptible to race condition, as it accesses `work_queues` without acquiring the corresponding lock.

Reply:
`condvars` lets the worker threads sleep without wasting any CPU time until there is work available for them to do -- `condvars[i]->wait(l)` pauses the thread until `convars[i]->notify_all()` is called on some other thread -- while the thread is asleep, the lock `l` is unlocked, but once it wakes up again it aquires `l` once more.

`done()` has a minor data race, but it's a read-only operation. Between jobs being added with `add_job` and finishing, `done()` will never `true`. If there are many jobs being added and finishing simultaneously, `done` might erroneously return `true`. `wait` waits for some job to complete, or for 100ms to pass, whichever happens first.

I should add these as comments :)

Reply:
Race condition invokes undefined behavior. Nothing, not even reading, is safe during a race.

But more importantly, I don't think we should have `done()` and `wait()`. I guess that the intention is signal to the main threading calling `open_as` when to wake up and recheck if the ID to be opened is no longer being closed. But

* Opening a ID that is just being closed is not a common scenario.
* The signal is not accurate. It only signals that one ID has been closed, not which ID.
* Each background thread has to call `job_done.notify_all()` after each closing. Not only does each call cost CPU cycles, it reduces parallelism since several threads may call it at the same time and have to wait for others.

In a nutshell, this design incurs a great cost, for a not-so-common scenario, and the main thread still has to recheck many times. Too much cost, too little gain.

Reply:
Good catch on the undefined behavior. I've been doing a bit too much platform-locked stuff so I forgot the CPP generic memory model. Even platform-locked to x86 this code may sometimes fail due to memory reordering effects in `empty()` -- so the body of the `if` should be in a lock guard regardless.

> Opening an ID that is just being closed is not a common scenario

Many directories in `find -type f` will fit this profile, with the current length of time deleting a `FileBase` takes.

> The signal is not accurate. It only signals that one ID has been closed, not which ID.

This is correct, but the signal *will* always trigger when the correct ID is closed, so it's just the pull-based version of this communication protocol. To wait on a particular ID, we'd have to have the thread pool know a lot more about the jobs being given to it. That *might* be a good idea, but given how the data race in `done()` snuck in, I think adding code complexity to the thread pool is risky.

> Each background thread has to call job_done.notify_all() after each closing. Not only does each call cost CPU cycles, it reduces parallelism since several threads may call it at the same time and have to wait for others.

Using something like condition variables will waste a lot less CPU time than periodic polling, and when an extended wait isn't necessary the condition variable will return quickly.

Reply:
It is best if you could do a real world benchmark on which approach is better (and how many threads are appropriate). If you insist on current design, though, I won't object.